test/images/stub-device-plugin/server.go

```diff
@@ -0,0 +1,316 @@
+	}, nil
+}
+
+// Register registers the device plugin for the given resourceName with Kubelet.
+func (s *StubDevicePlugin) Register(kubeletEndpoint, resourceName string) error {
+	conn, err := grpc.Dial(kubeletEndpoint, grpc.WithInsecure(),
+		grpc.WithDialer(func(addr string, timeout time.Duration) (net.Conn, error) {
+			return net.DialTimeout("unix", addr, timeout)
+		}))
+	defer conn.Close()
+	if err != nil {
+		return err
+	}
+	client := pluginapi.NewRegistrationClient(conn)
+	reqt := &pluginapi.RegisterRequest{
+		Version:      pluginapi.Version,
+		Endpoint:     path.Base(s.endpoint),
+		ResourceName: resourceName,
+	}
+
+	_, err = client.Register(context.Background(), reqt)
+	if err != nil {
+		return fmt.Errorf("Cannot register stub worker to kubelet service: %v", err)
+	}
+	return nil
+}
+
+// ListAndWatch lists devices and update that list according to the Update call
+func (s *StubDevicePlugin) ListAndWatch(emtpy *pluginapi.Empty, server pluginapi.DevicePlugin_ListAndWatchServer) error {
+	for {
+		resp := new(pluginapi.ListAndWatchResponse)
+		for _, dev := range s.devices {
+			resp.Devices = append(resp.Devices, &pluginapi.Device{
+				ID:     dev.ID,
+				Health: dev.Health,
+			})
+		}
+		log.Printf("ListAndWatch: send devices %v", resp)
+
+		if err := server.Send(resp); err != nil {
+			log.Printf("device-plugin: cannot update device states: %v\n", err)
+			return err
+		}
+
+		time.Sleep(5 * time.Second)
+	}
+}
+
+// Allocate which return list of devices.
+func (s *StubDevicePlugin) Allocate(ctx context.Context, r *pluginapi.AllocateRequest) (*pluginapi.AllocateResponse, error) {
+	var response pluginapi.AllocateResponse
+
+	for _, id := range r.DevicesIDs {
+		dev, exists := s.devices[id]
+		if !exists {
+			return nil, fmt.Errorf("Invalid allocation request with non-existing device %s", id)
+		}
+		if dev.Health != pluginapi.Healthy {
+			return nil, fmt.Errorf("Invalid allocation request with unhealthy device %s", id)
+		}
+		response.Devices = append(response.Devices, &pluginapi.DeviceSpec{
+			HostPath:      path.Join(s.watcherPath, dev.ID),
+			ContainerPath: path.Join(s.watcherPath, dev.ID),
+			Permissions:   "mrw",
+		})
+
+		response.Mounts = append(response.Mounts, &pluginapi.Mount{
+			ContainerPath: path.Join(s.watcherPath, dev.ID),
+			HostPath:      path.Join(s.watcherPath, dev.ID),
+		})
+	}
+
+	log.Printf("Allocate response: %+v", response)
+	return &response, nil
+}
+
+// Start will start grpc server and register to kubelet
+func (s *StubDevicePlugin) Start() error {
+	// Start grpc server
+	log.Printf("StubDevicePlugin start at: %s", s.endpoint)
+	err := s.cleanup()
+	if err != nil {
+		return err
+	}
+
+	socket, err := net.Listen("unix", s.endpoint)
+	if err != nil {
+		return err
+	}
+
+	s.grpcServer = grpc.NewServer()
+	pluginapi.RegisterDevicePluginServer(s.grpcServer, s)
+
+	go s.grpcServer.Serve(socket)
+
+	// Wait till the grpcServer is ready to serve services.
+	for {
+		s.mutex.Lock()
```

Comment:
Not sure what this lock is supposed to do ?

Reply:
It is just for race condition here.

Reply:
What race are you trying to prevent ?

Reply:
Oh, thanks! It seems no race condition here, i will remove it.